docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+# Support build and deployment of application container image using odo
+
+## Abstract
+Add a new command (verb) to build a production-like/slim container image for the project and deploy the built image on the target platform.
+
+There is an existing proposal for adding on outer-loop information (including support for multiple strategies for building and deploying the projects) to Devfile 2.1.0: 
+https://github.com/devfile/kubernetes-api/issues/49 
+
+It would be useful to start the design/development of a simpler version of `odo deploy` with devfile 2.0.0 that covers:
+- single build strategy - Dockerfile built using `buildah`
+- single deployment strategy - Kubernetes manifest deployed using `kubectl`.
+
+## Motivation
+`odo` is limited to development inner-loop for a project and there is no support for outer-loop - build a slim/production container image for your application and deploy the build container. It would be very useful for developers to be able to try inner-loop and then transition over to the outer-loop using odo. The outer-loop information could be provided by the application stack (devfile) to avoid developers having to worry about these aspects. 
+
+`odo deploy` can be a good way to assure the developer that their application can be built and deployed successfully using the build/deploy guidance that comes from devfile.
+
+It is not meant to replace GitOps/pipeline-based deployments for governed environments (test, stage, production). However, both `odo deploy` and `odo pipelines` must honour the build and deployment information provided by the application stack (devfile).
+
+## User Flow
+
+This command will allow a user to perform inner-loop and then test the outer-loop, so the application is truly ready for checking into git and for pipelines to take over. Here's how a typical application development flow might look like: 
+
+User flow: 
+1. `odo create <component> <mycomponent>` - This initializes odo component in the current directory.
+1. Edit the project source code to develop the application.
+1. `odo URL create` - This stores URL information (host, port etc.) for accessing the application on the cluster (if not done already).
+1. `odo push` - This runs the application source code using inner-loop instructions from devfile.
+1. Validate the running application is working as intended.
+1. Iterate over steps 2 and beyond (as needed). 
+1. `odo deploy` - This would build a new clean image and deploy it to the target cluster using outer-loop instructions from devfile and user-provided arguments.
```

Comment:
Would this be deploying the application "on the side" to the inner-loop so after this stage is complete the application will be running in two instances one in inner-loop mode and the other in outer-loop mode? Or is this more like a transition and the inner-loop run will be deleted?

I'm assuming that it is the first case as that would be more helpful, but it will create additional challenges like preventing name collisions when deployed to the same namespace.

Reply:
Yes, this would be the first case with there being 2 deployments, the inner-loop and then the outer-loop, each being independent of each other.  We would be looking at appending a specific value  to the names of the created objects, ie `-deploy` to ensure there are no name clashes with the inner-loop.  The value to be appended is still under discussion.

Reply:
Don't want to go too deep into the implementation details. But we need to be careful about this. Modifying resource names might have consequences. For example, k8s `Service` name maps to a name in internal k8s DNS. If we change its name without the user being awaited other services might not be able to connect to the deployed app.